# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieServiceChild.cpp
# Commit: 750e0796c1ad
# Full Hash: 750e0796c1ad0a93d26428fefe4419af324371a4
# Author: Paul Zuehlcke <pbz@mozilla.com>
# Date: 2023-09-07 21:34:02
# Regressor Bug: 1839918
# File Overlap Count: 3
# Description:
#   Bug 1839918 - Create SetCookies variant for document.cookie on PWindowGlobal. r=bvandersloot,cookie-reviewers,valentin
#   
#   Using PWindowGlobal for SetCookies calls from content to parent process means we can get
#   the BrowsingContext the cookies are set for. The current implementation in PCookieService
#   drops this relationship.
# ==============================================================================

diff -r 3949d9b400d4 -r 750e0796c1ad netwerk/cookie/CookieServiceChild.cpp
--- a/netwerk/cookie/CookieServiceChild.cpp	Thu Sep 07 08:25:55 2023 +0000
+++ b/netwerk/cookie/CookieServiceChild.cpp	Thu Sep 07 08:25:55 2023 +0000
@@ -33,6 +33,7 @@
 #include "mozilla/TimeStamp.h"
 #include "ThirdPartyUtil.h"
 #include "nsIConsoleReportCollector.h"
+#include "mozilla/dom/WindowGlobalChild.h"
 
 using namespace mozilla::ipc;
 
@@ -440,7 +441,16 @@
     cookiesToSend.AppendElement(cookie->ToIPC());
 
     // Asynchronously call the parent.
-    SendSetCookies(baseDomain, attrs, documentURI, false, cookiesToSend);
+    dom::WindowGlobalChild* windowGlobalChild =
+        aDocument->GetWindowGlobalChild();
+
+    // If there is no WindowGlobalChild fall back to PCookieService SetCookies.
+    if (NS_WARN_IF(!windowGlobalChild)) {
+      SendSetCookies(baseDomain, attrs, documentURI, false, cookiesToSend);
+      return NS_OK;
+    }
+    windowGlobalChild->SendSetCookies(baseDomain, attrs, documentURI, false,
+                                      cookiesToSend);
   }
 
   return NS_OK;